add options for setup of vue app instance

agentschmitt opened the pull request:
add options to the vue-plugin to allow setup of vue app instance. e.g. for registration of components, plugins...
many frameworks require this type of registration and cant be used otherwise.

usage would look like this:
```
    const render = new VuePlugin<Schemes, AreaExtra>({
        setupVue3: (app) => {
            app.use(i18n); // see https://vue-i18n.intlify.dev/guide/
            app.use(vuetify); // see https://vuetifyjs.com/en/getting-started/installation/
        }
    });
```

i had to make two different methods for vue2 and vue3 to get correct types exported.
so for vue2 it would be setupVue2 ;)

Ni55aN commented:
is it possible to combine these options into a version-agnostic one?

agentschmitt commented:
I had some deeper look into the setup logic for vue2 and found out that return of the vue instance doesnt work for setup use cases in vue2.
So i think i need to have two different setup method for vue2 and vue3.

For example setup of vue-i18n (8.x) for vue2 in main.ts would look this:
```
// for vue2 plugins can be registered globally
Vue.use(VueI18n); 

 // but the i18n instance need to be added to the vue instance
const i18n = new VueI18n({  locale: 'en', messages: { en: { message: { hello: 'hello world' } } } })

new Vue({
  render: (h) => h(App),
  i18n
}).$mount('#app')
```

In case of retejs it would be required to add the i18n instance to the vue instances that are created.
But it would not work doing this with something like
```
setupVue2: (app) => {
   app.$options.i18n = i18n;
}
```

So i think we need to change the setupVue2 method to set some options that can be added on creation of the vue instance internally in the plugin directly.

setupVue2: () => {
  return { i18n };
}

agentschmitt commented:
i updated the PR so it works for vue2 and vue3. usage would look like this.

```
// Vue2
const render = new VuePlugin<Schemes, AreaExtra>({ 
  setupVue2: () => { return { i18n } }
})

// Vue3
    const render = new VuePlugin<Schemes, AreaExtra>({
        setupVue3: (app) => { app.use(i18n);  }
    });
```

Ni55aN commented:
why not make it more generic, where the `setup` middleware just takes care of passing the necessary properties and waiting for the application instance?

Vue 2
```ts
const render = new VuePlugin<Schemes, AreaExtra>({ 
  setup: (context) => {
    const app = new Vue({ ...context, router })
    
    return app
  }
})
```
Vue 3
```ts
const render = new VuePlugin<Schemes, AreaExtra>({ 
  setup: (context) => {
    const app = createApp({ ...context })
    
    app.use(router)
    return app 
  }
})
```
I'm not sure if this will work since I didn't have a chance to test it

agentschmitt commented:
the setup would not look as simple as your sample above.
we would need to take care of many setup parts that are done inside the vue plugin currently.
for example passing the payload, passing the seed, call delegates for onRendered

Vue2
```
const render = new VuePlugin<Schemes, AreaExtra>({ 
  setup: (context) => {
    const app = new Vue({
      props: ['payload'],
      render(h) {
        return h(component, { props: { ...this.payload, seed: Math.random() }, ref: 'child' })
      },
      mounted() {
        onRendered()
      },
      updated() {
        onRendered()
      }
    })
    
    return app
  }
})
```

```
const render = new VuePlugin<Schemes, AreaExtra>({ 
  setup: (context) => {
    const app = createApp({
      render() {
        // @ts-ignore
        return h(component, { ...state.value, seed: Math.random() })
      },
      mounted() {
        onRendered()
      },
      updated() {
        onRendered()
      }
    })

    return app 
  }
})


```

agentschmitt commented:
> why not make it more generic, where the `setup` middleware just takes care of passing the necessary properties and waiting for the application instance?

i took this approach now and pass the whole setup stuff with as context parameter to the setup function.
so we need to create the app/vue instance ourself but can then adjust the creation as we like.

```
    const render = new VuePlugin<Schemes, AreaExtra>({
        setup: (context) => {
            const app = createApp(context);
            app.use(i18n);
            app.use(vuetify);
            return app;
        }
    });
```

Ni55aN commented:
thanks, I'll check this out as soon as I setup CI workflows for PRs (e.g. based on your suggested changes I can check project build and e2e tests for different stacks and versions)

Ni55aN commented:
I've setup CI for testing and ran tests based on your changes and [modified Rete Kit](https://github.com/retejs/rete-kit/tree/test/vue-plugin-pr-79) with usage of `i18n` via `setup` property:
- [CI regression tests Vue 2](https://github.com/retejs/rete-qa/actions/runs/7438277512)
- [CI regression tests Vue 3](https://github.com/retejs/rete-qa/actions/runs/7438277131)
- [CI tests Vue 2 with `i18n`](https://github.com/retejs/rete-qa/actions/runs/7438093506)
- [CI tests Vue 3 with `i18n`](https://github.com/retejs/rete-qa/actions/runs/7438094740)

Basically, it verifies that these changes work well (at least for the scenarios covered by the tests).

The only thing that, in my opinion, requires attention is the types (`(context: object) => object`). I'll investgate this later. Now the types are shared for the entire package (so they `/vue2` also uses these types), so we need a solution on how they can be specified  for different versions of Vue

Ni55aN commented on src/index.ts:
```diff
@@ -21,6 +21,18 @@ type Requires<Schemes extends BaseSchemes> =
+   *  @param [context] to be used for createApp({ ...context }) or new Vue({ ...context })
+   *  @returns app / vue instance.
+   */
+  setup?: (context: object) => object;
```
```ts
  setup?: (context: Context) => Instance;
```

these types can be defined in `./vuecompat/types.ts`:

```ts

import type VueNamespace from 'vue'

type Vue = typeof VueNamespace

export type Context = Vue extends { createApp: (arg: infer U) => any }
  ? U
  : (Vue extends { new(options: infer U): any } ? U : any)

export type Instance = Vue extends { createApp: (arg: any) => infer U }
  ? U
  : (Vue extends { new(options: any): infer U } ? U : any)
```

This way, it can infer an options and return type depending on Vue.js version

agentschmitt replied:
i squashed my commits to one "feat" commit for the new configuration feature.
i also added your dynamic typings as "refactor" commit, looks fine for me.

Ni55aN reviewed (changes requested):
Additionally, can you rewrite commit mesages according to [Semantic Commit Messages](https://gist.github.com/alpteo/e93d754e5e09907c6362c4230fb66f87)? (it's required for semantic release)
